### PR TITLE
Remove panic logic from services

### PR DESCRIPTION
**Pull request by benpankow**
Addresses #213. We now directly handle errors instead of panicking and relying on the errors middleware to defer/recover. This lets us maintain the flow of requests through the API and ensures that middleware can correctly handle the request on the way out.

- [x] TODO: Remove [errors middleware](https://github.com/HackIllinois/api/blob/master/common/middleware/error.go), which should no longer be needed.

### CODE REVIEW

**Comment by ASankaran** [comment]
You might have a plan for this already, but the logging logic from the errors middleware should be moved into the WriteError function somehow. Otherwise debugging issues in prod will be near impossible.

File: common/errors/errors.go [hunk 1]
```diff
@@ -14,3 +21,38 @@ type ApiError struct {
...
+
+	if err != nil {
+		fmt.Printf("Failed to marshal error for id: %v\n", id)
+		return
+	}
+
+	fmt.Printf("ERROR: %v\n", string(error_log_message))
+}
+
+// Writes the given error to the passed HTTP response
```
Comment (bstrauch24) [inline common/errors/errors.go]: Code is descriptive enough; don't need this comment

File: common/errors/errors.go [hunk 2]
```diff
@@ -14,3 +21,38 @@ type ApiError struct {
...
+	Error interface{}
+}
+
+func LogError(id string, error_message interface{}) {
+	log_entry := ErrorLogEntry{
+		ID:    id,
+		Error: error_message,
+	}
+
+	error_log_message, err := json.MarshalIndent(log_entry, "", "    ")
```
Comment (bstrauch24) [inline common/errors/errors.go]: Hard to tell what this code is doing without looking up what `MarshalIndent` means; maybe put the "tab" string in a constant called `INDENT`?

File: services/auth/controller/controller.go [hunk 3]
```diff
@@ -41,13 +41,15 @@ func Authorize(w http.ResponseWriter, r *http.Request) {
 	oauth_provider, err := service.GetOAuthProvider(provider)
 
 	if err != nil {
-		panic(errors.MalformedRequestError(err.Error(), "Invalid OAuth provider."))
+		errors.WriteError(w, r, errors.MalformedRequestError(err.Error(), "Invalid OAuth provider."))
```
Comment (bstrauch24) [inline services/auth/controller/controller.go]: Hopefully you used a regex for all of these!

File: common/apiserver/apiserver.go [hunk 4]
```diff
@@ -19,7 +19,6 @@ func StartServer(address string, router *mux.Router, name string, initialize fun
 		return err
 	}
 
-	router.Use(middleware.ErrorMiddleware)
```
Comment (bstrauch24) [inline common/apiserver/apiserver.go]: Do we still need the code in `api/common/middleware/error.go`?

Comment (bstrauch24) [reply common/apiserver/apiserver.go]: Never mind, I think you deleted it. GitHub doesn't show that very well

**Pull request merged**